### PR TITLE
std: Add basic smoke test for net functionality

### PR DESCRIPTION
**Pull request by LemonBoy**
Barely enough testing to make sure `net` doesn't keep on breaking every damn time.
The async tests are broken as exepcted, patches welcome.

Open questions:
* What the everloving fuck is a `socket_t`? Are we minting our own faux-libc types? Why not `SocketHandle`? 
* Why the hell does the `IPAddress` struct hold a port? It's an ip address!
* Why is `IPAddress` a wrapper over `sockaddr_in` rather than something saner like a fixed-size array of octets? The number of casts and pointer reinterpretations needed to fill/access them is making (me and) the stage1 compiler sick. Zig currently barfs when trying to parse an ip at comptime!
* Why the IPv6 parser is doing syscalls in the parsing routine?? The ipv6 scope should be kept as a literal and looked up only when needed (since `Ip6Address` is a `sockaddr_in6` you _always_ need to resolve it right away...)

### CODE REVIEW

File: lib/std/os/bits/windows.zig [hunk 1]
```diff
@@ -256,6 +256,41 @@ pub const POLLERR = ws2_32.POLLERR;
 pub const POLLHUP = ws2_32.POLLHUP;
 pub const POLLNVAL = ws2_32.POLLNVAL;
 
+pub const SOL_SOCKET = 0xffff;
+
+pub const SO_DEBUG = 0x0001;
```
Comment (daurnimator) [inline lib/std/os/bits/windows.zig:261]: Should these come from `ws2_32`?

Comment (LemonBoy) [reply lib/std/os/bits/windows.zig]: Yup, I keep losing track of the many places where bits go.

**Comment by LemonBoy** [comment]
We absolutely need to separate the socket handles from the regular file ones, I've just discovered that you have to call `closesocket` on the former rather than using `close`.

x-ref #6600.

**Comment by daurnimator** [comment]
>  I've just discovered that you have to call `closesocket` on the former rather than using `close`.

Only if you go via win32; not if you use the native api.

**Comment by LemonBoy** [comment]
> Only if you go via win32; not if you use the native api.

Isn't winsock the native api?

**Comment by daurnimator** [comment]
> Isn't winsock the native api?

the native api is AFD.

**Comment by LemonBoy** [comment]
> the native api is AFD.

Good lord, the undocumented kernel driver that's not supposed to be used from usermode?

**Comment by daurnimator** [comment]
> undocumented

> not supposed to be used from usermode

According to microsoft the entire NTAPI is "not supposed to be used from usermode" and is mostly undocumented.
Winsocket has gained a lot of cruft over the years; using AFD directly is not only more efficient, but has better APIs.
Note that projects like libuv (and hence the entire node js javascript ecosystem) in several cases use AFD directly: Microsoft won't be breaking is time soon.

**Comment by andrewrk** [comment]
>     * What the everloving fuck is a `socket_t`? Are we minting our own faux-libc types? Why not `SocketHandle`?

It comes from WASI. I think this a good example of something that would benefit from your points in #6600.

>     * Why the hell does the `IPAddress` struct hold a port? It's an ip address!

idk, it made sense at the time but I'm not prepared to defend the decision. The std.net code is young, untested, unvetted, and certainly will be reworked, probably more than once.

> [other questions]

Zig is a project that has slowly gained quality and robustness by iterating and reworking - not by getting the design right the first time. The short answer to all this stuff is: good points, let's improve things. It's all work-in-progress.

**Pull request merged**